src/backend/distributed/transaction/remote_transaction.c

```diff
@@ -820,6 +845,269 @@ CoordinatedRemoteTransactionsAbort(void)
+		if (transaction->transactionFailed && !transaction->transactionRecovering)
+		{
+			continue;
+		}
+
+		FinishRemoteTransactionSavepointRollback(connection, subId);
+	}
+}
+
+
+/*
+ * StartRemoteTransactionSavepointBegin initiates SAVEPOINT command for the given
+ * subtransaction id in a non-blocking manner.
+ */
+static void
+StartRemoteTransactionSavepointBegin(MultiConnection *connection, SubTransactionId subId)
+{
+	StringInfo savepointCommand = makeStringInfo();
+	appendStringInfo(savepointCommand, "SAVEPOINT savepoint_%u", subId);
+
+	if (!SendRemoteCommand(connection, savepointCommand->data))
+	{
+		ReportConnectionError(connection, WARNING);
+		MarkRemoteTransactionFailed(connection, true);
+	}
+}
+
+
+/*
+ * FinishRemoteTransactionSavepointBegin finishes the work
+ * StartRemoteTransactionSavepointBegin initiated. It blocks if necessary (i.e.
+ * if PQisBusy() would return true).
+ */
+static void
+FinishRemoteTransactionSavepointBegin(MultiConnection *connection, SubTransactionId subId)
+{
+	PGresult *result = GetRemoteCommandResult(connection, true);
+	if (!IsResponseOK(result))
+	{
+		ReportResultError(connection, result, WARNING);
+		MarkRemoteTransactionFailed(connection, true);
+	}
+
+	PQclear(result);
+	ForgetResults(connection);
+}
+
+
+/*
+ * StartRemoteTransactionSavepointRelease initiates RELEASE SAVEPOINT command for
+ * the given subtransaction id in a non-blocking manner.
+ */
+static void
+StartRemoteTransactionSavepointRelease(MultiConnection *connection,
+									   SubTransactionId subId)
+{
+	StringInfo savepointCommand = makeStringInfo();
+	appendStringInfo(savepointCommand, "RELEASE SAVEPOINT savepoint_%u", subId);
+
+	if (!SendRemoteCommand(connection, savepointCommand->data))
+	{
+		ReportConnectionError(connection, WARNING);
+		MarkRemoteTransactionFailed(connection, true);
+	}
+}
+
+
+/*
+ * FinishRemoteTransactionSavepointRelease finishes the work
+ * StartRemoteTransactionSavepointRelease initiated. It blocks if necessary (i.e.
+ * if PQisBusy() would return true).
+ */
+static void
+FinishRemoteTransactionSavepointRelease(MultiConnection *connection,
+										SubTransactionId subId)
+{
+	PGresult *result = GetRemoteCommandResult(connection, true);
+	if (!IsResponseOK(result))
+	{
+		ReportResultError(connection, result, WARNING);
+		MarkRemoteTransactionFailed(connection, true);
+	}
+
+	PQclear(result);
+	ForgetResults(connection);
+}
+
+
+/*
+ * StartRemoteTransactionSavepointRollback initiates ROLLBACK TO SAVEPOINT command
+ * for the given subtransaction id in a non-blocking manner.
+ */
+static void
+StartRemoteTransactionSavepointRollback(MultiConnection *connection,
+										SubTransactionId subId)
+{
+	StringInfo savepointCommand = makeStringInfo();
+	appendStringInfo(savepointCommand, "ROLLBACK TO SAVEPOINT savepoint_%u", subId);
+
+	if (!SendRemoteCommand(connection, savepointCommand->data))
+	{
+		ReportConnectionError(connection, WARNING);
+		MarkRemoteTransactionFailed(connection, true);
```

Comment:
Based on our internal discussion, I added conditions for this at begin/release but not at rollback.